packages/base/src/Render.js

```diff
@@ -0,0 +1,174 @@
+const eventProvider = new EventProvider();
+
+const invalidatedWebComponents = new RenderQueue(); // Queue for invalidated web components
+
+let renderTaskPromise,
+	renderTaskPromiseResolve;
+
+let mutationObserverTimer;
+
+let queuePromise;
+
+
+/**
+ * Schedules a render task (if not already scheduled) to render the component
+ *
+ * @param webComponent
+ * @returns {Promise}
+ */
+const renderDeferred = async webComponent => {
+	// Enqueue the web component
+	invalidatedWebComponents.add(webComponent);
+
+	// Schedule a rendering task
+	await scheduleRenderTask();
+};
+
+/**
+ * Renders a component synchronously and adds it to the registry of rendered components
+ *
+ * @param webComponent
+ */
+const renderImmediately = webComponent => {
+	eventProvider.fireEvent("beforeComponentRender", webComponent);
+	registeredElements.add(webComponent);
+	webComponent._render();
+};
+
+/**
+ * Cancels the rendering of a component, if awaiting to be rendered, and removes it from the registry of rendered components
+ *
+ * @param webComponent
+ */
+const cancelRender = webComponent => {
+	invalidatedWebComponents.remove(webComponent);
+	registeredElements.delete(webComponent);
+};
+
+/**
+ * Schedules a rendering task, if not scheduled already
+ */
+const scheduleRenderTask = async () => {
+	if (!queuePromise) {
+		queuePromise = new Promise(resolve => {
+			window.requestAnimationFrame(() => {
+				// Render all components in the queue
+
+				// console.log(`--------------------RENDER TASK START------------------------------`); // eslint-disable-line
+				invalidatedWebComponents.process(renderImmediately);
+				// console.log(`--------------------RENDER TASK END------------------------------`); // eslint-disable-line
+
+				// Resolve the promise so that callers of renderDeferred can continue
+				queuePromise = null;
+				resolve();
+
+				// Wait for Mutation observer before the render task is considered finished
+				if (!mutationObserverTimer) {
+					mutationObserverTimer = setTimeout(() => {
+						mutationObserverTimer = undefined;
+						if (invalidatedWebComponents.isEmpty()) {
+							_resolveTaskPromise();
+						}
+					}, 200);
+				}
+			});
+		});
+	}
+
+	await queuePromise;
+};
+
+/**
+ * return a promise that will be resolved once all invalidated web components are rendered
+ */
+const whenDOMUpdated = () => {
+	if (renderTaskPromise) {
+		return renderTaskPromise;
+	}
+
+	renderTaskPromise = new Promise(resolve => {
+		renderTaskPromiseResolve = resolve;
+		window.requestAnimationFrame(() => {
+			if (invalidatedWebComponents.isEmpty()) {
+				renderTaskPromise = undefined;
+				resolve();
+			}
+		});
+	});
+
+	return renderTaskPromise;
+};
+
+const whenAllCustomElementsAreDefined = () => {
+	const definedPromises = getAllRegisteredTags().map(tag => customElements.whenDefined(tag));
+	return Promise.all(definedPromises);
+};
+
+const whenFinished = async () => {
+	await whenAllCustomElementsAreDefined();
+	await whenDOMUpdated();
```

Comment:
not related to this change, but I think we should improve this with a loop, as the DOM Update promise might result in more undefined custom elements.